lib/workflows/map_reduce.rb

```diff
@@ -0,0 +1,115 @@
+require "services"
+require "milemarker"
+require "sidekiq/batch"
+require "tmpdir"
+require "sidekiq_jobs"
+require "workflow_component"
+
+module Workflows
+  # Gets data (one record per line) from a data source, splits it into chunks,
+  # then runs a job on each chunk (the "mapper"); when the mapping steps
+  # complete, runs a callback (the "reducer") with the output
+  # from the mapper.
+  #
+  # Jobs are run on files containing the given records_per_job lines (default 10,000)
+  class MapReduce
+    class Callback
+      def on_success(_status, options)
+        reducer = Object.const_get(options["component_class"])
+        reducer_params = Jobs.symbolize(options["params"])
+        reducer.new(**reducer_params).run
+
+        # Don't delete intermediate files for now to aid in debugging.
+        # In the future: consider removing by default, but disabling that if a debug flag is present.
+        # FileUtils.remove_entry(options["working_directory"])
+      end
+    end
+
+    def initialize(
+      working_directory: default_working_directory,
+      records_per_job: 10000,
+      components: {},
```

Comment:
I don't love that components is a hash, but we are validating it, and doing it this way (vs. dynamically defining more attributes/methods/whatever) seems better. I had tried a version where there was a `WorkflowSpec` that encapsulated the components, but IMO it ended up increasing verbosity without increasing clarity.